File: web/nextjs/lib/api/waf-lab.ts
import { apiClient } from "@/lib/api/client";

export type WAFLabRunStatus = "queued" | "running" | "completed" | "failed" | "cancelled";

export interface WAFLabCategoryScore {
  total: number;
  secure_blocked: number;
  open_leaked: number;
  false_positives: number;
}

export interface WAFLabScore {
  attacks_total: number;
  secure_blocked: number;
  secure_expected: number;
  open_leaked: number;
  open_expected_leak: number;
  false_positives: number;
  not_blocked_expected: number;
  by_category: Record<string, WAFLabCategoryScore>;
  suspicious: boolean;
  suspicious_reason?: string;
  bola_note?: string;
}

export interface WAFLabRun {
  id: string;
  agent_id: string;
  task_id?: string | null;
  org_id: string;
  status: WAFLabRunStatus;
  mode: string;
  secure_host: string;
  open_host: string;
  origin_upstream: string;
  policy_id: string;
  manage_dns: string;
  dns_zone: string;
  attack_set: string;
  instruction?: string | null;
  wslproxy_base_url: string;
  score?: WAFLabScore | null;
  error_message?: string | null;
  requested_by?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface WAFLabStep {
  id: string;
  run_id: string;
  phase: string;
  status: string;
  message: string;
  started_at?: string | null;
  ended_at?: string | null;
  created_at: string;
}

export interface WAFLabResult {
  id: string;
  run_id: string;
  host_role: "secure" | "open" | string;
  host: string;
  attack_id: string;
  attack_name: string;
  category: string;
  method: string;
  path: string;
  payload?: string;
  status_code: number;
  blocked: boolean;
  expect_block: boolean;
  waf_rule?: string;
  waf_violation?: string;
  support_id?: string;
  latency_ms: number;
  verdict: string;
  notes?: string;
  created_at: string;
}

export interface CreateWAFLabRunRequest {
  agent_id: string;
  task_id?: string;
  wslproxy_base_url: string;
  secure_host: string;
  open_host: string;
  origin_upstream?: string;
  policy_id?: string;
  mode: string;
  manage_dns?: string;
  dns_zone?: string;
  attack_set?: string;
  instruction?: string;
}

export interface PaginatedWAFLabRuns {
  data: WAFLabRun[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export async function wafLabStatus() {
  const { data } = await apiClient.get<Record<string, unknown>>("/waf-lab/status");
  return data;
}

export async function listWafLabRuns(page = 1, perPage = 10) {
  const { data } = await apiClient.get<PaginatedWAFLabRuns>("/waf-lab/runs", {
    params: { page, per_page: perPage },
  });
  return data;
}

export async function createWafLabRun(body: CreateWAFLabRunRequest) {
  const { data } = await apiClient.post<WAFLabRun>("/waf-lab/runs", body);
  return data;
}

export async function getWafLabRun(runID: string) {
  const { data } = await apiClient.get<WAFLabRun>(`/waf-lab/runs/${runID}`);
  return data;
}

export async function listWafLabSteps(runID: string) {
  const { data } = await apiClient.get<WAFLabStep[]>(`/waf-lab/runs/${runID}/steps`);
  return data;
}

export async function listWafLabResults(runID: string) {
  const { data } = await apiClient.get<WAFLabResult[]>(`/waf-lab/runs/${runID}/results`);
  return data;
}

export async function cancelWafLabRun(runID: string) {
  const { data } = await apiClient.post<WAFLabRun>(`/waf-lab/runs/${runID}/cancel`);
  return data;
}

File: server/internal/model/waf_lab.go
package model

import (
	"time"

	"github.com/google/uuid"
)

const AgentNameWAFLab = "WAF Efficacy Lab"

// CreateWAFLabRunRequest is the launch payload for a WAF Efficacy Lab run.
type CreateWAFLabRunRequest struct {
	AgentID         uuid.UUID  `json:"agent_id" validate:"required"`
	TaskID          *uuid.UUID `json:"task_id"`
	WSLProxyBaseURL string     `json:"wslproxy_base_url" validate:"required"`
	SecureHost      string     `json:"secure_host" validate:"required"`
	OpenHost        string     `json:"open_host" validate:"required"`
	OriginUpstream  string     `json:"origin_upstream"`
	PolicyID        string     `json:"policy_id"`
	Mode            string     `json:"mode" validate:"required,oneof=provision_and_test provision_only test_only"`
	ManageDNS       string     `json:"manage_dns" validate:"omitempty,oneof=off cloudflare"`
	DNSZone         string     `json:"dns_zone"`
	AttackSet       string     `json:"attack_set" validate:"omitempty,oneof=full owasp_core modern_api stages_only"`
	Instruction     string     `json:"instruction" validate:"omitempty,max=4000"`
}

// WAFLabRun is one efficacy-lab execution.
type WAFLabRun struct {
	ID              uuid.UUID  `json:"id"`
	AgentID         uuid.UUID  `json:"agent_id"`
	TaskID          *uuid.UUID `json:"task_id"`
	OrgID           uuid.UUID  `json:"org_id"`
	Status          string     `json:"status"` // queued, running, completed, failed, cancelled
	Mode            string     `json:"mode"`
	SecureHost      string     `json:"secure_host"`
	OpenHost        string     `json:"open_host"`
	OriginUpstream  string     `json:"origin_upstream"`
	PolicyID        string     `json:"policy_id"`
	ManageDNS       string     `json:"manage_dns"`
	DNSZone         string     `json:"dns_zone"`
	AttackSet       string     `json:"attack_set"`
	Instruction     *string    `json:"instruction,omitempty"`
	WSLProxyBaseURL string     `json:"wslproxy_base_url"`
	Score           *WAFLabScore `json:"score,omitempty"`
	ErrorMessage    *string    `json:"error_message,omitempty"`
	RequestedBy     *uuid.UUID `json:"requested_by"`
	StartedAt       *time.Time `json:"started_at"`
	CompletedAt     *time.Time `json:"completed_at"`
	CreatedAt       time.Time  `json:"created_at"`
	UpdatedAt       time.Time  `json:"updated_at"`
}

// WAFLabStep is one phase of a lab run.
type WAFLabStep struct {
	ID        uuid.UUID  `json:"id"`
	RunID     uuid.UUID  `json:"run_id"`
	Phase     string     `json:"phase"` // preflight, rules_policy, hosts, dns, efficacy
	Status    string     `json:"status"` // pending, running, completed, failed, skipped
	Message   string     `json:"message"`
	StartedAt *time.Time `json:"started_at"`
	EndedAt   *time.Time `json:"ended_at"`
	CreatedAt time.Time  `json:"created_at"`
}

// WAFLabResult is one attack × host outcome.
type WAFLabResult struct {
	ID           uuid.UUID `json:"id"`
	RunID        uuid.UUID `json:"run_id"`
	HostRole     string    `json:"host_role"` // secure | open
	Host         string    `json:"host"`
	AttackID     string    `json:"attack_id"`
	AttackName   string    `json:"attack_name"`
	Category     string    `json:"category"`
	Method       string    `json:"method"`
	Path         string    `json:"path"`
	Payload      string    `json:"payload,omitempty"`
	StatusCode   int       `json:"status_code"`
	Blocked      bool      `json:"blocked"`
	ExpectBlock  bool      `json:"expect_block"`
	WAFRule      string    `json:"waf_rule,omitempty"`
	WAFViolation string    `json:"waf_violation,omitempty"`
	SupportID    string    `json:"support_id,omitempty"`
	LatencyMS    int       `json:"latency_ms"`
	Verdict      string    `json:"verdict"` // blocked, leaked, false_positive, not_blocked_expected, error
	Notes        string    `json:"notes,omitempty"`
	CreatedAt    time.Time `json:"created_at"`
}

// WAFLabScore summarises a before/after matrix.
type WAFLabScore struct {
	AttacksTotal       int            `json:"attacks_total"`
	SecureBlocked      int            `json:"secure_blocked"`
	SecureExpected     int            `json:"secure_expected"`
	OpenLeaked         int            `json:"open_leaked"`
	OpenExpectedLeak   int            `json:"open_expected_leak"`
	FalsePositives     int            `json:"false_positives"`
	NotBlockedExpected int            `json:"not_blocked_expected"` // e.g. BOLA
	ByCategory         map[string]WAFLabCategoryScore `json:"by_category"`
	Suspicious         bool           `json:"suspicious"`
	SuspiciousReason   string         `json:"suspicious_reason,omitempty"`
	BOLANote           string         `json:"bola_note,omitempty"`
}

// WAFLabCategoryScore is per-category coverage.
type WAFLabCategoryScore struct {
	Total          int `json:"total"`
	SecureBlocked  int `json:"secure_blocked"`
	OpenLeaked     int `json:"open_leaked"`
	FalsePositives int `json:"false_positives"`
}
